tests/suite/graceful_recovery_test.go

```diff
@@ -0,0 +1,316 @@
+	)).ToNot(HaveOccurred())
+}
+
+func waitForContainerRestart(ngfPodName, containerName string, currentRestartCount int) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutConfig.RequestTimeout)
+	defer cancel()
+
+	//nolint:nilerr
+	return wait.PollUntilContextCancel(
+		ctx,
+		500*time.Millisecond,
+		true, /* poll immediately */
+		func(_ context.Context) (bool, error) {
+			restartCount, err := getContainerRestartCount(containerName, ngfPodName)
+			if err != nil {
+				return false, nil
+			}
+
+			return restartCount == currentRestartCount+1, nil
+		},
+	)
+}
+
+func waitForWorkingTraffic() error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutConfig.RequestTimeout)
+	defer cancel()
+
+	//nolint:nilerr
+	return wait.PollUntilContextCancel(
+		ctx,
+		500*time.Millisecond,
+		true, /* poll immediately */
+		func(_ context.Context) (bool, error) {
+			if err := expectRequestToSucceed(teaURL, address, "URI: /tea"); err != nil {
+				return false, nil
+			}
+			if err := expectRequestToSucceed(coffeeURL, address, "URI: /coffee"); err != nil {
+				return false, nil
+			}
+			return true, nil
+		},
+	)
+}
+
+func waitForFailingTraffic() error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutConfig.RequestTimeout)
+	defer cancel()
+
+	//nolint:nilerr
+	return wait.PollUntilContextCancel(
+		ctx,
+		500*time.Millisecond,
+		true, /* poll immediately */
+		func(_ context.Context) (bool, error) {
+			if err := expectRequestToFail(teaURL, address, "URI: /tea"); err != nil {
+				return false, nil
+			}
+			if err := expectRequestToFail(coffeeURL, address, "URI: /coffee"); err != nil {
+				return false, nil
+			}
+			return true, nil
+		},
+	)
+}
+
+func expectRequestToSucceed(appURL, address string, responseBodyMessage string) error {
+	status, body, err := framework.Get(appURL, address, timeoutConfig.RequestTimeout)
+	if status != http.StatusOK {
+		return errors.New("http status was not 200")
+	}
+
+	if !strings.Contains(body, responseBodyMessage) {
+		return errors.New("expected response body to contain correct body message")
+	}
+
+	return err
+}
+
+func expectRequestToFail(appURL, address string, responseBodyMessage string) error {
+	status, body, err := framework.Get(appURL, address, timeoutConfig.RequestTimeout)
+	if status != 0 {
+		return errors.New("expected http status to be 0")
+	}
+
+	if strings.Contains(body, responseBodyMessage) {
+		return errors.New("expected response body to not contain correct body message")
+	}
+
+	if err == nil {
+		return errors.New("expected request to error")
+	}
+
+	return nil
+}
+
+func checkContainerLogsForErrors(ngfPodName string) {
+	sinceSeconds := int64(15)
+	logs, err := resourceManager.GetPodLogs(
+		ngfNamespace,
+		ngfPodName,
+		&core.PodLogOptions{Container: nginxContainerName, SinceSeconds: &sinceSeconds},
+	)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(logs).ToNot(ContainSubstring("emerg"), logs)
```

Comment:
With NGINX, you can use `[emerg]` as a search target, as NGINX logs that string in the error log.

emerg is log level. Other log level exist such as `error`. An error log line will be logged with `[error]` string in it.

We're interested in error, crit, alert, or emerg lol levels (https://nginx.org/en/docs/ngx_core_module.html#error_log)

Could you also search for error, crit and alert?